### ISSUE TITLE
[BUG] Concurrent access in maps when the element is delete

### ISSUE THREAD

**Issue by yunwei37** [eunomia-bpf/bpftime#320]
**Describe the bug**

The similar situation like https://github.com/iovisor/ubpf/issues/45#issuecomment-755073584 may exist.

**Comment by yunwei37** [comment]
Verified: the pattern from iovisor/ubpf#45 (comment 755073584) does exist in bpftime's user-space shared-memory maps today.

What happens:

- Per-map ops take the per-handler spinlock only for the duration of a single helper call (`map_lookup_elem` / `map_update_elem` / `map_delete_elem` in `runtime/src/handler/map_handler.cpp`). The lock is released as soon as the helper returns, and `elem_lookup` hands the raw pointer to the program: the variable-size hash map returns `&itr->second[0]` (`runtime/src/bpf_map/userspace/var_hash_map.cpp`), the LRU map returns `&itr->second.value[0]`, and the fixed-size hash map returns a bucket slot of the preallocated buffer.
- After `bpf_map_lookup_elem()` returns, the program therefore holds an open pointer while another process sharing the map (or another thread of the same process, since helpers run on the caller's threads) can delete the element. `map_impl.erase(itr)` destroys the shared-memory byte vectors and returns their memory to the segment allocator, so the program's in-place update afterwards is a use-after-free whose window spans many instructions. LRU eviction triggered by an unrelated `update` frees the same way, without any explicit delete. The kernel avoids this with RCU - element memory is not freed/reused before the program exits - and bpftime currently has no equivalent reclamation guarantee.
- The open-addressing fixed hash (`runtime/src/bpf_map/bpftime_hash_map.hpp`) never frees and only marks the bucket deleted, but a later `elem_update` may refill that bucket between the lookup and the program's in-place write, so the write silently merges into a different key's value.
- The common iterate-with-`get_next_key` + `delete` pattern amplifies both variants.

Not exposed the same way: array maps (preallocated, no delete), kernel-backed maps (the syscall path copies values), ringbuf. Per-CPU hash map runs fully unlocked (`should_lock == false`), so concurrent helpers on different CPUs race on the same `boost::unordered_map` container without any synchronization as well.

Side finding from this verification: per-CPU hash `elem_delete` fills `[begin, begin + cpu * value_size)` with 0, but it should be `[cpu * value_size, (cpu + 1) * value_size)` otherwise it wipes the wrong CPUs' slices and never clears its own (`runtime/src/bpf_map/userspace/per_cpu_hash_map.cpp`).

Possible directions:

1. Document the constraint: a pointer returned by `bpf_map_lookup_elem` is only valid until any other map operation on the same map. Cheapest, but it is a real divergence from kernel semantics that programs legitimately rely on.
2. Deferred reclamation: on delete/eviction, move the storage blocks to a per-handler retire list and recycle them only after a grace period has been observed (epoch-based, similar to liburcu but small enough to hand-roll per handler). That makes var hash / LRU erasure safe without changing per-op cost much; fixed-hash bucket reuse can go through the same deferred free list.
3. For LRU specifically, eviction could move entries to a tombstone state first and reap them later on the same path.

Keeping this open as a tracked bug.

AI-generated response; a maintainer will review and follow up later
